From 7a93cfd6d5a75eb63976dcafd399eb4177685f11 Mon Sep 17 00:00:00 2001
From: George Cherian <george.cherian@marvell.com>
Date: Wed, 19 May 2021 17:34:28 +0530
Subject: [PATCH 1538/1921] octeontx2-npa: Make the freeing IRQ vector fix
 complete

In case of mailbox timeout in npa_register_mbox_intr(), make sure to
free the allocated mbox interrupt before returning -EPROBE_DEFER.
Update the err variable and take the common error path.

Change-Id: I19da5b7bb81be518e94a1e25b743962146948808
Fixes: 977196c5ad8e3 ("octeontx2-npa: Fix freeing IRQ vectors")
Signed-off-by: George Cherian <george.cherian@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/52370
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Sunil Kovvuri Goutham <sgoutham@marvell.com>
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 drivers/soc/marvell/octeontx2-npa/npa.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/soc/marvell/octeontx2-npa/npa.c b/drivers/soc/marvell/octeontx2-npa/npa.c
index 7a7750c88c64..610d2a9cd874 100644
--- a/drivers/soc/marvell/octeontx2-npa/npa.c
+++ b/drivers/soc/marvell/octeontx2-npa/npa.c
@@ -1220,8 +1220,8 @@ static int npa_register_mbox_intr(struct npa_dev_t *npa_pf_dev, bool probe_af)
 	if (err) {
 		dev_warn(&npa_pf_dev->pdev->dev,
 			 "AF not responding to mailbox, deferring probe\n");
-		otx2_disable_afpf_mbox_intr(npa_pf_dev);
-		return -EPROBE_DEFER;
+		err = -EPROBE_DEFER;
+		goto err_free_intr;
 	}
 
 	mutex_lock(&npa_pf_dev->lock);
-- 
2.31.1

